src/mono/wasm/runtime/diagnostics.ts

```diff
@@ -0,0 +1,142 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { Module } from "./imports";
+import cwraps from "./cwraps";
+import type { EventPipeSessionOptions } from "./types";
+import type { VoidPtr } from "./types/emscripten";
+import * as memory from "./memory";
+
+const sizeOfInt32 = 4;
+
+export type EventPipeSessionID = bigint;
+type EventPipeSessionIDImpl = number;
+
+/// An EventPipe session object represents a single diagnostic tracing session that is collecting
+/// events from the runtime and managed libraries.  There may be multiple active sessions at the same time.
+/// Each session subscribes to a number of providers and will collect events from the time that start() is called, until stop() is called.
+/// Upon completion the session saves the events to a file on the VFS.
+/// The data can then be retrieved as Blob or as a data URI (prefer Blob).
+export interface EventPipeSession {
+    // session ID for debugging logging only
+    get sessionID(): EventPipeSessionID;
+    start(): void;
+    stop(): void;
+    getTraceBlob(): Blob;
+}
+
+// internal session state of the JS instance
+enum State {
+    Initialized,
+    Started,
+    Done,
+}
+
+function withStackAlloc<TRes> (bytesWanted: number, f : (ptr: VoidPtr) => TRes): TRes {
```

Comment:
Nice, please move it to `memory.ts` 
@kg do you think that allocating the callback closure is too expensive for general usage ?

Reply:
yes, but it doesn't seem to be used in hot paths so this is probably fine

Reply:
https://github.com/kg/runtime/commit/2c62ddf2c541f515714532afd53bc7b944a871eb#diff-d320e7f4617dada664d0674a4bd8ee2a1e8eae9118f7b4320112e70d20ebfb49R106-R109 is an example of one way to avoid the closure by adding userdata arguments instead

Reply:
ok, I added declarations for up to 7 UD arguments.  

But it's kind of hard to avoid creating a closure in TS/JS because even module imports are closures.

Reply:
makes sense, thank you. I rewrote my use of `withStackAlloc` to use a helper function for the callback. (I don't expect EP sessions to be created that aggressively, but I want to get the muscle memory for doing things right for other stuff that might be on the hot path)